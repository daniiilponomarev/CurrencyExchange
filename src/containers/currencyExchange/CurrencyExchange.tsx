import React from "react";
import { useSelector, useDispatch } from "react-redux";
import Box from "@mui/material/Box";
import Container from "@mui/material/Container";
import Card from "@mui/material/Card";
import CardContent from "@mui/material/CardContent";
import CardActions from "@mui/material/CardActions";
import Button from "@mui/material/Button";
import CurrencyExchangeIcon from "@mui/icons-material/CurrencyExchange";
import ChangeCircleTwoToneIcon from "@mui/icons-material/ChangeCircleTwoTone";
import IconButton from "@mui/material/IconButton";
import Typography from "@mui/material/Typography";

import { GAPS } from "../../constants";
import {
  selectBoughtCurrency,
  selectSoldCurrency,
  exchangeCurrencies,
} from "./currencyExchangeSlice";
import { BuyCurrency, CurrencyRate, SellCurrency } from "./components";
import {
  cardActionsStyle,
  cardContentStyle,
  cardStyle,
  currenciesSelectsBoxStyle,
  headersContainerStyle,
} from "./CurrencyExchangeStyles";

export const CurrencyExchange = () => {
  const soldCurrency = useSelector(selectSoldCurrency);
  const boughtCurrency = useSelector(selectBoughtCurrency);
  const dispatch = useDispatch();
  const handleExchangeCurrencies = () => dispatch(exchangeCurrencies());

  return (
    <Box sx={{ width: "50%", minWidth: "450px" }}>
      <Card variant="outlined" sx={cardStyle}>
        <CardContent sx={cardContentStyle}>
          <Container sx={headersContainerStyle}>
            <CurrencyExchangeIcon fontSize="large" color="primary" />
            <Typography variant="h4">Currency Exchange</Typography>
          </Container>
          <Typography variant="h5" sx={{ textAlign: "center", mb: GAPS.xs }}>
            Buy {boughtCurrency}
          </Typography>
          <Box sx={currenciesSelectsBoxStyle}>
            <CurrencyRate />
            <SellCurrency />
            <IconButton
              aria-label="exchange currencies"
              color="primary"
              onClick={handleExchangeCurrencies}
            >
              <ChangeCircleTwoToneIcon />
            </IconButton>
            <BuyCurrency />
          </Box>
        </CardContent>

        <CardActions sx={cardActionsStyle}>
          <Button size="small" variant="contained">
            {"Sell"} {soldCurrency} {"for"} {boughtCurrency}
          </Button>
        </CardActions>
      </Card>
    </Box>
  );
};
