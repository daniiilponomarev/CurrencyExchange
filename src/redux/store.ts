import { configureStore } from "@reduxjs/toolkit";
import currencyExchangeReducer, {
  CurrencyExchangeState,
} from "../containers/currencyExchange/currencyExchangeSlice";
import balancesReducer, { BalancesState } from "./balancesSlice";

export type State = {
  currencyExchange: CurrencyExchangeState;
  balances: BalancesState;
};

export default configureStore({
  reducer: {
    currencyExchange: currencyExchangeReducer,
    balances: balancesReducer,
  },
});
