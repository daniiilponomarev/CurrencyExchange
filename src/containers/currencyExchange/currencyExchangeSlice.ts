import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { State } from "../../redux/store";
import { CurrencyName } from "../../types";

export type CurrencyExchangeState = {
  soldCurrency: CurrencyName;
  boughtCurrency: CurrencyName;
  rate: number | null;
  soldCurrencyAmount: number;
  boughtCurrencyAmount: number;
};

const initialState: CurrencyExchangeState = {
  soldCurrency: "USD",
  boughtCurrency: "EUR",
  rate: null,
  soldCurrencyAmount: 0,
  boughtCurrencyAmount: 0,
};

export const slice = createSlice({
  name: "currencyExchange",
  initialState,
  reducers: {
    setSoldCurrency: (state, action: PayloadAction<CurrencyName>) => {
      state.soldCurrency = action.payload;
    },
    setBoughtCurrency: (state, action: PayloadAction<CurrencyName>) => {
      state.boughtCurrency = action.payload;
    },
    setSoldCurrencyAmount: (state, action: PayloadAction<number>) => {
      const value = action.payload;
      const newBoughtCurrencyAmount = value * (state.rate || 1);

      state.soldCurrencyAmount = value;
      state.boughtCurrencyAmount = newBoughtCurrencyAmount;
    },
    setBoughtCurrencyAmount: (state, action: PayloadAction<number>) => {
      const value = action.payload;
      const newSoldCurrencyAmount = value / (state.rate || 1);

      state.boughtCurrencyAmount = value;
      state.soldCurrencyAmount = newSoldCurrencyAmount;
    },
    exchangeCurrencies: (state) => {
      const soldCurrencyTemp = state.soldCurrency;
      state.soldCurrency = state.boughtCurrency;
      state.boughtCurrency = soldCurrencyTemp;
    },
    setRate: (state, action: PayloadAction<number | null>) => {
      state.rate = action.payload;
    },
  },
});

export const {
  setSoldCurrency,
  setBoughtCurrency,
  exchangeCurrencies,
  setRate,
  setSoldCurrencyAmount,
  setBoughtCurrencyAmount,
} = slice.actions;

export const selectSoldCurrency = (state: State) =>
  state.currencyExchange.soldCurrency;
export const selectBoughtCurrency = (state: State) =>
  state.currencyExchange.boughtCurrency;
export const selectSoldCurrencyAmount = (state: State) =>
  state.currencyExchange.soldCurrencyAmount;
export const selectBoughtCurrencyAmount = (state: State) =>
  state.currencyExchange.boughtCurrencyAmount;
export const selectRate = (state: State) => state.currencyExchange.rate;

export default slice.reducer;
