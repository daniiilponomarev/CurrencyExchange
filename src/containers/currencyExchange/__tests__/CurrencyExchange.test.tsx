import React from "react";
import { Provider } from "react-redux";
import { QueryClient, QueryClientProvider } from "react-query";
import {
  render,
  RenderResult,
  screen,
  fireEvent,
} from "@testing-library/react";

import { CurrencyExchange } from "../CurrencyExchange";
import store from "../../../redux/store";
import { setRate, setSoldCurrencyAmount } from "../currencyExchangeSlice";

jest.mock("../components", () => ({
  BuyCurrency: () => "BuyCurrency",
  CurrencyRate: () => "CurrencyRate",
  SellCurrency: () => "SellCurrency",
}));
const queryClient = new QueryClient();

const renderCurrencyExchange = (): RenderResult =>
  render(
    <Provider store={store}>
      <QueryClientProvider client={queryClient}>
        <CurrencyExchange />
      </QueryClientProvider>
    </Provider>
  );
describe("<CurrencyExchange />", () => {
  test("should match snapshot", () => {
    const component = renderCurrencyExchange();

    expect(component).toMatchSnapshot();
  });

  test("should exchange currencies", () => {
    renderCurrencyExchange();

    const button = screen.getByTestId("CurrencyExchange-exchange-button");

    const currencyExchangeOld = store.getState().currencyExchange;
    fireEvent.click(button);
    const currencyExchangeNew = store.getState().currencyExchange;

    expect(currencyExchangeNew.boughtCurrency).toEqual(
      currencyExchangeOld.soldCurrency
    );
    expect(currencyExchangeNew.soldCurrency).toEqual(
      currencyExchangeOld.boughtCurrency
    );
  });

  test("should submit exchanging currencies", () => {
    renderCurrencyExchange();

    store.dispatch(setRate(2));
    store.dispatch(setSoldCurrencyAmount(10));

    const balancesOld = store.getState().balances;
    console.log(store.getState().balances);
    console.log(store.getState().currencyExchange);
    const button = screen.getByTestId("CurrencyExchange-submit-button");
    fireEvent.click(button);

    const currencyExchangeNew = store.getState().balances;
    expect(currencyExchangeNew.EUR).toEqual(balancesOld.EUR - 10);
    expect(currencyExchangeNew.USD).toEqual(balancesOld.USD + 20);
  });
});
