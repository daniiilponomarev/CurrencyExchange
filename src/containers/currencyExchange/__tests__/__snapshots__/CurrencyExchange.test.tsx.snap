// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrencyExchange /> should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiBox-root css-5uqiy3"
      >
        <div
          class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-13svwm0-MuiPaper-root-MuiCard-root"
        >
          <div
            class="MuiCardContent-root css-gpkk0i-MuiCardContent-root"
          >
            <div
              class="MuiContainer-root MuiContainer-maxWidthLg css-12fnwcq-MuiContainer-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeLarge css-bsqp9w-MuiSvgIcon-root"
                data-testid="CurrencyExchangeIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12.89 11.1c-1.78-.59-2.64-.96-2.64-1.9 0-1.02 1.11-1.39 1.81-1.39 1.31 0 1.79.99 1.9 1.34l1.58-.67c-.15-.45-.82-1.92-2.54-2.24V5h-2v1.26c-2.48.56-2.49 2.86-2.49 2.96 0 2.27 2.25 2.91 3.35 3.31 1.58.56 2.28 1.07 2.28 2.03 0 1.13-1.05 1.61-1.98 1.61-1.82 0-2.34-1.87-2.4-2.09l-1.66.67c.63 2.19 2.28 2.78 2.9 2.96V19h2v-1.24c.4-.09 2.9-.59 2.9-3.22 0-1.39-.61-2.61-3.01-3.44zM3 21H1v-6h6v2H4.52c1.61 2.41 4.36 4 7.48 4 4.97 0 9-4.03 9-9h2c0 6.08-4.92 11-11 11-3.72 0-7.01-1.85-9-4.67V21zm-2-9C1 5.92 5.92 1 12 1c3.72 0 7.01 1.85 9 4.67V3h2v6h-6V7h2.48C17.87 4.59 15.12 3 12 3c-4.97 0-9 4.03-9 9H1z"
                />
              </svg>
              <h4
                class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
              >
                Currency Exchange
              </h4>
            </div>
            <h5
              class="MuiTypography-root MuiTypography-h5 css-1mau8h2-MuiTypography-root"
            >
              Buy 
              EUR
            </h5>
            CurrencyRate
            <div
              class="MuiBox-root css-h8123p"
            >
              SellCurrency
              <button
                aria-label="exchange currencies"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-15rynab-MuiButtonBase-root-MuiIconButton-root"
                data-testid="CurrencyExchange-exchange-button"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ChangeCircleTwoToneIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m.06 9.34v2.14c-.92.02-1.84-.31-2.54-1.01-1.12-1.12-1.3-2.8-.59-4.13l-1.1-1.1c-1.28 1.94-1.07 4.59.64 6.29.97.98 2.25 1.47 3.53 1.47h.06v2l2.83-2.83-2.83-2.83zm3.48-4.88c-.99-.99-2.3-1.46-3.6-1.45V5L9.11 7.83l2.83 2.83V8.51H12c.9 0 1.79.34 2.48 1.02 1.12 1.12 1.3 2.8.59 4.13l1.1 1.1c1.28-1.94 1.07-4.59-.63-6.3z"
                    opacity=".3"
                  />
                  <path
                    d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm.06 11.34v2.14c-.92.02-1.84-.31-2.54-1.01-1.12-1.12-1.3-2.8-.59-4.13l-1.1-1.1c-1.28 1.94-1.07 4.59.64 6.29.97.98 2.25 1.47 3.53 1.47h.06v2l2.83-2.83-2.83-2.83zm3.48-4.88c-.99-.99-2.3-1.46-3.6-1.45V5L9.11 7.83l2.83 2.83V8.51H12c.9 0 1.79.34 2.48 1.02 1.12 1.12 1.3 2.8.59 4.13l1.1 1.1c1.28-1.94 1.07-4.59-.63-6.3z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              BuyCurrency
            </div>
          </div>
          <div
            class="MuiCardActions-root MuiCardActions-spacing css-umyros-MuiCardActions-root"
          >
            <button
              class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root  css-jh47zj-MuiButtonBase-root-MuiButton-root"
              data-testid="CurrencyExchange-submit-button"
              tabindex="0"
              type="button"
            >
              Sell
               
              USD
               
              for
               
              EUR
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiBox-root css-5uqiy3"
    >
      <div
        class="MuiPaper-root MuiPaper-outlined MuiPaper-rounded MuiCard-root css-13svwm0-MuiPaper-root-MuiCard-root"
      >
        <div
          class="MuiCardContent-root css-gpkk0i-MuiCardContent-root"
        >
          <div
            class="MuiContainer-root MuiContainer-maxWidthLg css-12fnwcq-MuiContainer-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeLarge css-bsqp9w-MuiSvgIcon-root"
              data-testid="CurrencyExchangeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12.89 11.1c-1.78-.59-2.64-.96-2.64-1.9 0-1.02 1.11-1.39 1.81-1.39 1.31 0 1.79.99 1.9 1.34l1.58-.67c-.15-.45-.82-1.92-2.54-2.24V5h-2v1.26c-2.48.56-2.49 2.86-2.49 2.96 0 2.27 2.25 2.91 3.35 3.31 1.58.56 2.28 1.07 2.28 2.03 0 1.13-1.05 1.61-1.98 1.61-1.82 0-2.34-1.87-2.4-2.09l-1.66.67c.63 2.19 2.28 2.78 2.9 2.96V19h2v-1.24c.4-.09 2.9-.59 2.9-3.22 0-1.39-.61-2.61-3.01-3.44zM3 21H1v-6h6v2H4.52c1.61 2.41 4.36 4 7.48 4 4.97 0 9-4.03 9-9h2c0 6.08-4.92 11-11 11-3.72 0-7.01-1.85-9-4.67V21zm-2-9C1 5.92 5.92 1 12 1c3.72 0 7.01 1.85 9 4.67V3h2v6h-6V7h2.48C17.87 4.59 15.12 3 12 3c-4.97 0-9 4.03-9 9H1z"
              />
            </svg>
            <h4
              class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
            >
              Currency Exchange
            </h4>
          </div>
          <h5
            class="MuiTypography-root MuiTypography-h5 css-1mau8h2-MuiTypography-root"
          >
            Buy 
            EUR
          </h5>
          CurrencyRate
          <div
            class="MuiBox-root css-h8123p"
          >
            SellCurrency
            <button
              aria-label="exchange currencies"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeLarge css-15rynab-MuiButtonBase-root-MuiIconButton-root"
              data-testid="CurrencyExchange-exchange-button"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ChangeCircleTwoToneIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m.06 9.34v2.14c-.92.02-1.84-.31-2.54-1.01-1.12-1.12-1.3-2.8-.59-4.13l-1.1-1.1c-1.28 1.94-1.07 4.59.64 6.29.97.98 2.25 1.47 3.53 1.47h.06v2l2.83-2.83-2.83-2.83zm3.48-4.88c-.99-.99-2.3-1.46-3.6-1.45V5L9.11 7.83l2.83 2.83V8.51H12c.9 0 1.79.34 2.48 1.02 1.12 1.12 1.3 2.8.59 4.13l1.1 1.1c1.28-1.94 1.07-4.59-.63-6.3z"
                  opacity=".3"
                />
                <path
                  d="M12 4c4.41 0 8 3.59 8 8s-3.59 8-8 8-8-3.59-8-8 3.59-8 8-8m0-2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm.06 11.34v2.14c-.92.02-1.84-.31-2.54-1.01-1.12-1.12-1.3-2.8-.59-4.13l-1.1-1.1c-1.28 1.94-1.07 4.59.64 6.29.97.98 2.25 1.47 3.53 1.47h.06v2l2.83-2.83-2.83-2.83zm3.48-4.88c-.99-.99-2.3-1.46-3.6-1.45V5L9.11 7.83l2.83 2.83V8.51H12c.9 0 1.79.34 2.48 1.02 1.12 1.12 1.3 2.8.59 4.13l1.1 1.1c1.28-1.94 1.07-4.59-.63-6.3z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            BuyCurrency
          </div>
        </div>
        <div
          class="MuiCardActions-root MuiCardActions-spacing css-umyros-MuiCardActions-root"
        >
          <button
            class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButtonBase-root  css-jh47zj-MuiButtonBase-root-MuiButton-root"
            data-testid="CurrencyExchange-submit-button"
            tabindex="0"
            type="button"
          >
            Sell
             
            USD
             
            for
             
            EUR
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
